src/borg/helpers.py

```diff
@@ -451,22 +451,45 @@ def __init__(self, fallback=None):
         # Value to return from match function when none of the patterns match.
         self.fallback = fallback
 
+        # optimizations
+        self._path_full_patterns = {}  # full path -> return value
+
     def empty(self):
-        return not len(self._items)
+        return not len(self._items) and not len(self._path_full_patterns)
+
+    def _add(self, pattern, value):
+        if isinstance(pattern, PathFullPattern):
+            key = pattern.pattern  # full, normalized path
+            self._path_full_patterns[key] = value
+        else:
+            self._items.append((pattern, value))
 
     def add(self, patterns, value):
         """Add list of patterns to internal list. The given value is returned from the match function when one of the
         given patterns matches.
         """
-        self._items.extend((i, value) for i in patterns)
+        for pattern in patterns:
+            self._add(pattern, value)
 
     def add_inclexcl(self, patterns):
         """Add list of patterns (of type InclExclPattern) to internal list. The patterns ptype member is returned from
         the match function when one of the given patterns matches.
         """
-        self._items.extend(patterns)
+        for pattern, pattern_type in patterns:
+            self._add(pattern, pattern_type)
 
     def match(self, path):
+        # do a fast lookup for full path matches:
+        # TODO: path = normalized(path), see normalized_wrapper
+        #       pulling this out of N .match() calls and only doing it once is also a good perf. optimization
+        # XXX we do not count such matches, do we need to?
```

Comment:
looks like "no", we only count matches for include patterns coming from args.paths, so we can warn if some never matched, but not for other patterns.